sql/2024/04/inaccurate-sizes-attribute-impact.sql

```diff
@@ -0,0 +1,106 @@
+# HTTP Archive query to measure impact of inaccurate sizes attributes per <img> for WordPress sites.
+#
+# WPP Research, Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+CREATE TEMPORARY FUNCTION
+  getSrcsetSizesAccuracy(payload STRING)
+  RETURNS ARRAY<STRUCT<sizesAbsoluteError INT64,
+  sizesRelativeError FLOAT64,
+  wDescriptorAbsoluteError INT64,
+  wDescriptorRelativeError FLOAT64,
+  actualSizesEstimatedWastedLoadedPixels INT64,
+  actualSizesEstimatedWastedLoadedBytes FLOAT64,
+  wastedLoadedPercent FLOAT64>>
+  LANGUAGE js AS '''
+try {
+  var $ = JSON.parse(payload);
+  var responsiveImages = JSON.parse($._responsive_images);
+  responsiveImages = responsiveImages['responsive-images'];
+  return responsiveImages.map(({
+    sizesAbsoluteError,
+    sizesRelativeError,
+    wDescriptorAbsoluteError,
+    wDescriptorRelativeError,
+    idealSizesSelectedResourceEstimatedPixels,
+    actualSizesEstimatedWastedLoadedPixels,
+    actualSizesEstimatedWastedLoadedBytes
+  }) => {
+    let wastedLoadedPercent;
+    if ( idealSizesSelectedResourceEstimatedPixels > 0 ) {
+      wastedLoadedPercent = actualSizesEstimatedWastedLoadedPixels / idealSizesSelectedResourceEstimatedPixels;
+    } else {
+      wastedLoadedPercent = null;
+    }
+    return {
+      sizesAbsoluteError,
+      sizesRelativeError,
+      wDescriptorAbsoluteError,
+      wDescriptorRelativeError,
+      actualSizesEstimatedWastedLoadedPixels,
+      actualSizesEstimatedWastedLoadedBytes,
+      wastedLoadedPercent
+    };
+  }
+);
+} catch (e) {
+  return [];
+}
+''';
+
+CREATE TEMPORARY FUNCTION IS_CMS(technologies ARRAY<STRUCT<technology STRING, categories ARRAY<STRING>, info ARRAY<STRING>>>, cms STRING, version STRING) RETURNS BOOL AS (
+  EXISTS(
+    SELECT * FROM UNNEST(technologies) AS technology, UNNEST(technology.info) AS info
+    WHERE technology.technology = cms
+    AND (
+      version = ""
+      OR ENDS_WITH(version, ".x") AND (STARTS_WITH(info, RTRIM(version, "x")) OR info = RTRIM(version, ".x"))
+      OR info = version
+    )
+  )
+);
+
+WITH wordpressSizesData AS (
+  SELECT
+    client,
+    image
+  FROM
+    `httparchive.all.pages`,
+    UNNEST(getSrcsetSizesAccuracy(payload)) AS image
```

Comment:
Since we're relying on the transparent logic from https://github.com/HTTPArchive/custom-metrics/blob/main/dist/responsive_images.js, we should probably use the `custom_metrics` column instead of `payload`. While the data is likely similar, the `custom_metrics` column is based on exactly this logic, while `payload` comes from the WebPageTest data, where we don't have insight how it is calculated - even though the custom metrics logic appears to replicate what WebPageTest does.